### PR TITLE
Make test attributes explicit

### PR DESCRIPTION
**Pull request by bf4**
Follows https://github.com/rails-api/active_model_serializers/pull/1982

### CODE REVIEW

File: lib/active_model_serializers/model.rb [hunk 1]
```diff
@@ -51,5 +59,43 @@ def self.lookup_ancestors
...
+    private
+
+    def _assign_attributes(attributes)
+      attributes.each do |k, v|
+        _assign_attribute(k, v)
+      end
+    end
+
+    def _assign_attribute(k, v)
+      fail UnknownAttributeError.new(self, k) unless respond_to?("#{k}=")
```
Comment (bf4) [inline lib/active_model_serializers/model.rb]: TODO: check vs. `attribute_names`

File: lib/active_model_serializers/model.rb [hunk 2]
```diff
@@ -51,5 +59,43 @@ def self.lookup_ancestors
...
+      public_send("#{k}=", v)
+    end
+
+    def persisted?
+      false
+    end
+
+    # Raised when unknown attributes are supplied via mass assignment.
+    class UnknownAttributeError < NoMethodError
```
Comment (bf4) [inline lib/active_model_serializers/model.rb]: copied from Rails 5 ActiveModel::AttributeAssignment

File: test/action_controller/adapter_selector_test.rb [hunk 3]
```diff
@@ -15,7 +15,7 @@ def render_using_adapter_override
         end
 
         def render_skipping_adapter
-          @profile = Profile.new(name: 'Name 1', description: 'Description 1', comments: 'Comments 1')
+          @profile = Profile.new(id: 'render_skipping_adapter_id', name: 'Name 1', description: 'Description 1', comments: 'Comments 1')
```
Comment (bf4) [inline test/action_controller/adapter_selector_test.rb]: probably a good thing that `id`, which is required, is now included

File: lib/active_model_serializers/model.rb [hunk 4]
```diff
@@ -3,42 +3,50 @@
 # serializable non-activerecord objects.
 module ActiveModelSerializers
   class Model
-    include ActiveModel::Model
     include ActiveModel::Serializers::JSON
+    include ActiveModel::Validations
+    include ActiveModel::Conversion
+    extend ActiveModel::Naming
+    extend ActiveModel::Translation
```
Comment (bf4) [inline lib/active_model_serializers/model.rb]: code adapted from ActiveModel::Model which is has good logic in it, but doesn't still has a few cases we want to handle differently

Comment (NullVoxPopuli) [inline lib/active_model_serializers/model.rb]: what are those use cases?

**Comment by NullVoxPopuli** [comment]
LGTM, ready for merge?

File: test/action_controller/serialization_test.rb [hunk 5]
```diff
@@ -135,7 +135,7 @@ def render_fragment_changed_object_with_relationship
           like = Like.new(id: 1, likeable: comment, time: 3.days.ago)
 
           generate_cached_serializer(like)
-          like.likable = comment2
+          like.likeable = comment2
```
Comment (bf4) [inline test/action_controller/serialization_test.rb]: typo!

File: test/fixtures/poro.rb [hunk 6]
```diff
@@ -3,23 +3,19 @@
...
-  # Convenience when not adding @attributes readers and writers
-  def method_missing(meth, *args)
-    if meth.to_s =~ /^(.*)=$/
-      attributes[Regexp.last_match(1).to_sym] = args[0]
-    elsif attributes.key?(meth)
-      attributes[meth]
-    else
-      super
-    end
+  # At this time, just for organization of intent
```
Comment (bf4) [inline test/fixtures/poro.rb]: probably good to promote to parent class once we add some additional behavior to the `attributes` method

File: test/active_model_serializers/model_test.rb [hunk 7]
```diff
@@ -18,5 +18,52 @@ def test_initialization_with_string_keys
...
+      assert_equal 1, instance.read_attribute_for_serialization(:one)
+
+      # Change via accessor
+      instance.one = :not_one
+
+      expected_attributes = { id: nil, one: :not_one, two: 2, three: 3 }.with_indifferent_access
+      assert_equal expected_attributes, instance.attributes
+      assert_equal :not_one, instance.one
+      assert_equal :not_one, instance.read_attribute_for_serialization(:one)
+    end
```
Comment (bf4) [inline test/active_model_serializers/model_test.rb]: is fixed https://github.com/rails-api/active_model_serializers/pull/1903 

ref https://github.com/rails-api/active_model_serializers/issues/1877

**Comment by bf4** [comment]
cc @richmolj

File: test/active_model_serializers/model_test.rb (hunk 7 above)
Comment (bf4) [inline test/active_model_serializers/model_test.rb]: bug ref https://github.com/rails-api/active_model_serializers/pull/1857#r75954677

File: lib/active_model_serializers/model.rb [hunk 8]
```diff
@@ -3,42 +3,50 @@
...
-    include ActiveModel::Model
     include ActiveModel::Serializers::JSON
+    include ActiveModel::Model
+
+    class_attribute :attribute_names
+    self.attribute_names = []
 
     def self.attributes(*names)
       attr_accessor(*names)
+      self.attribute_names = attribute_names | names.map(&:to_sym)
```
Comment (beauby) [inline lib/active_model_serializers/model.rb]:
```ruby
self.attribute-names |= names.map(&:to_sym)
```

File: lib/active_model_serializers/model.rb [hunk 9]
```diff
@@ -3,42 +3,50 @@
...
     def initialize(attributes = {})
-      @attributes = attributes && attributes.symbolize_keys
       @errors = ActiveModel::Errors.new(self)
       super
     end
 
     # Defaults to the downcased model name.
     def id
-      attributes.fetch(:id) { self.class.name.downcase }
+      @id ||= self.class.name && self.class.name.downcase
```
Comment (beauby) [inline lib/active_model_serializers/model.rb]: The `id` method should return a unique identifier for a resource, no? Not sure if providing such a default is more useful than dangerous.

File: lib/active_model_serializers/model.rb [hunk 10]
```diff
@@ -3,42 +3,50 @@
...
     end
 
     # Defaults to the downcased model name and updated_at
     def cache_key
-      attributes.fetch(:cache_key) { "#{self.class.name.downcase}/#{id}-#{updated_at.strftime('%Y%m%d%H%M%S%9N')}" }
+      ActiveSupport::Cache.expand_cache_key([
+        self.class.name && self.class.name.downcase,
```
Comment (beauby) [inline lib/active_model_serializers/model.rb]: How about defaulting to `self.class.object_id` for anonymous classes? That way you avoid colliding all anonymous classes in the cache.

File: lib/active_model_serializers/model.rb [hunk 11]
```diff
@@ -3,42 +3,50 @@
...
+        "#{id}-#{updated_at.strftime('%Y%m%d%H%M%S%9N')}"
+      ].compact)
     end
 
     # Defaults to the time the serializer file was modified.
     def updated_at
-      attributes.fetch(:updated_at) { File.mtime(__FILE__) }
+      defined?(@updated_at) ? @updated_at : File.mtime(__FILE__)
```
Comment (beauby) [inline lib/active_model_serializers/model.rb]: It is safe to do `@updated_at || File.mtime(__FILE__)`.

File: lib/active_model_serializers/model.rb (hunk 9 above)
Comment (bf4) [inline lib/active_model_serializers/model.rb]: @beauby I agree this is a somewhat problematic implementation.  Also, I think only JSON API requires an 'id'.

File: lib/active_model_serializers/model.rb (hunk 10 above)
Comment (bf4) [inline lib/active_model_serializers/model.rb]: I thought about that, or `hash`, but I'm not sure that's any better, as it's not really a unique identifier. We could hash the attributes, I guess.  If we just used `ActiveModel::Name` then it would blow up on the anonymous class and we could rely on it's exception message.

File: lib/active_model_serializers/model.rb (hunk 11 above)
Comment (bf4) [inline lib/active_model_serializers/model.rb]: @beauby yeah, but give ruby warning undefined ivar

File: lib/active_model_serializers/model.rb (hunk 9 above)
Comment (beauby) [inline lib/active_model_serializers/model.rb]: In that case, I'd suggest trying to read the `:id` attribute, and if none available, fall back to `object_id`.

File: lib/active_model_serializers/model.rb (hunk 10 above)
Comment (beauby) [inline lib/active_model_serializers/model.rb]: `self.class.object_id` **is** a unique identifier for the class. For the the instances, I don't see any better solution than hashing the attributes, which would suck because it would either be complicated or fail for hashes. I'd vote for no caching for those models.

File: lib/active_model_serializers/model.rb [hunk 12]
```diff
@@ -1,44 +1,48 @@
...
-    def read_attribute_for_serialization(key)
-      if key == :id || key == 'id'
-        attributes.fetch(key) { id }
-      else
-        attributes[key]
-      end
+    def attributes
+      attribute_names.each_with_object({}) do |attribute_name, result|
+        result[attribute_name] = public_send(attribute_name)
+      end.with_indifferent_access
```
Comment (bf4) [inline lib/active_model_serializers/model.rb]: `.freeze`

File: lib/active_model_serializers/model/caching.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+      included do
+        attr_writer :updated_at
+        attributes :id
+      end
+
+      # Defaults to the downcased model name and updated_at
+      def cache_key
+        ActiveSupport::Cache.expand_cache_key([
+          self.class.model_name.name.downcase,
+          "#{id}-#{updated_at.strftime('%Y%m%d%H%M%S%9N')}"
```
Comment (bf4) [inline lib/active_model_serializers/model/caching.rb]: This will blow up if `updated_at` isn't a 'Time', but that will only happen when users set updated_at, since the default is `File.mtime`

File: test/active_model_serializers/model_test.rb [hunk 14]
```diff
@@ -18,5 +27,59 @@ def test_initialization_with_string_keys
 
       assert_equal model_instance.read_attribute_for_serialization(:key), value
     end
+
+    def test_attributes_can_be_read_for_serialization
+      klass = subclass_without_caching_support do
```
Comment (bf4) [inline test/active_model_serializers/model_test.rb]: specifically, this prevents the `id` from being added

File: test/active_model_serializers/model_test.rb [hunk 15]
```diff
@@ -18,5 +27,59 @@ def test_initialization_with_string_keys
...
+      # Change via accessor
+      instance.one = :not_one
+
+      expected_attributes = { one: :not_one, two: 2, three: 3 }.with_indifferent_access
+      assert_equal expected_attributes, instance.attributes
+      assert_equal :not_one, instance.one
+      assert_equal :not_one, instance.read_attribute_for_serialization(:one)
+    end
+
+    def test_id_attribute_can_be_read_for_serialization
```
Comment (bf4) [inline test/active_model_serializers/model_test.rb]: no longer necessary now that the `read_attribute_for_serialization` method has been removed

File: test/active_model_serializers/model_test.rb (hunk 14 above)
Comment (bf4) [inline test/active_model_serializers/model_test.rb]: But, since `id` no no longer has special behavior, subclassing without caching methods isn't necessary here and can be moved to another PR, if we like

File: test/cache_test.rb [hunk 16]
```diff
@@ -34,6 +34,7 @@ class UncachedAuthor < Author
     end
 
     class Article < ::Model
+      attributes :title
       # To confirm error is raised when cache_key is not set and cache_key option not passed to cache
       undef_method :cache_key
```
Comment (bf4) [inline test/cache_test.rb]: seems like we want to make  a `NonCachingModel` poro base class for cases like this.

File: test/fixtures/poro.rb [hunk 17]
```diff
@@ -3,23 +3,24 @@
...
+
+  def self.associations(*names)
+    attr_accessor(*names)
+    self.association_names = association_names | names.map(&:to_sym)
+  end
+
+  def associations
+    association_names.each_with_object({}) do |association_name, result|
+      result[association_name] = public_send(association_name)
+    end.with_indifferent_access
```
Comment (bf4) [inline test/fixtures/poro.rb]: `.freeze` but this isn't actually called right now...

File: test/fixtures/poro.rb [hunk 18]
```diff
@@ -30,67 +31,74 @@ def respond_to_missing?(method_name, _include_private = false)
...
 end
-
 class ProfileSerializer < ActiveModel::Serializer
   attributes :name, :description
 
-  # TODO: is this used anywhere?
+  # ActiveModel::Serializer::OptionsTest#test_options_are_accessible:
+  # test/serializers/options_test.rb:12:in `test_options_are_accessible'
+  # test/serializers/options_test.rb:17:in `test_no_option_is_passed_in'
   def arguments_passed_in?
```
Comment (bf4) [inline test/fixtures/poro.rb]: TODO: move to `test/serializers/options_test.rb`

File: test/fixtures/poro.rb [hunk 19]
```diff
@@ -30,67 +31,74 @@ def respond_to_missing?(method_name, _include_private = false)
...
+  attributes :id
+  has_many :posts
+end
+
 class Comment < Model
+  attributes :body, :date
+  associations :post, :author, :likes
+
   # Uses a custom non-time-based cache key
   def cache_key
```
Comment (bf4) [inline test/fixtures/poro.rb]: still used?

File: test/fixtures/poro.rb [hunk 20]
```diff
@@ -30,67 +31,74 @@ def respond_to_missing?(method_name, _include_private = false)
...
-
-class ObjectTag < ActiveRecord::Base
-  belongs_to :poly_tag
-  belongs_to :taggable, polymorphic: true
-end
-
-class Picture < ActiveRecord::Base
-  belongs_to :imageable, polymorphic: true
-  has_many :object_tags, as: :taggable
+  def custom_options
```
Comment (bf4) [inline test/fixtures/poro.rb]: necessary?

File: test/fixtures/poro.rb [hunk 21]
```diff
@@ -103,143 +111,135 @@ def blog
...
-  # TODO: is this used anywhere?
+  # ActiveModel::Serializer::ArraySerializerTest#test_each_object_should_be_serialized_with_appropriate_serializer:
+  # test/collection_serializer_test.rb:43:in `test_each_object_should_be_serialized_with_appropriate_serializer'
+  # ActiveModel::Serializer::ArraySerializerTest#test_serializer_option_not_passed_to_each_serializer:
+  # test/collection_serializer_test.rb:49:in `test_serializer_option_not_passed_to_each_serializer'
+  # ActiveModel::Serializer::CollectionSerializerTest#test_each_object_should_be_serialized_with_appropriate_serializer:
+  # test/collection_serializer_test.rb:43:in `test_each_object_should_be_serialized_with_appropriate_serializer'
+  # ActiveModel::Serializer::CollectionSerializerTest#test_serializer_option_not_passed_to_each_serializer:
+  # test/collection_serializer_test.rb:49:in `test_serializer_option_not_passed_to_each_serializer'
   def custom_options
```
Comment (bf4) [inline test/fixtures/poro.rb]: TODO: move to `test/collection_serializer_test.rb`

File: test/fixtures/poro.rb [hunk 22]
```diff
@@ -103,143 +111,135 @@ def blog
...
-  has_one :blog, key: :site
+class Tag < Model
+  attributes :name
 end
 
+class VirtualValue < Model; end
 class VirtualValueSerializer < ActiveModel::Serializer
   attributes :id
-
   has_many :reviews, virtual_value: [{ type: 'reviews', id: '1' },
```
Comment (bf4) [inline test/fixtures/poro.rb]: I wonder if anyone actually uses this

Comment (NullVoxPopuli) [inline test/fixtures/poro.rb]: I think we should remove it. It's not as clean as defining a reviews method, or using the block form.

File: test/fixtures/poro.rb (hunk 20 above)
Comment (groyoh) [inline test/fixtures/poro.rb]: I guess it is used to test that options are passed from collection serializer -> serializer and serializer -> association_serializer but it could definitely benefit from creating a specific serializer within the file testing this case.

File: lib/active_model_serializers/model.rb [hunk 23]
```diff
@@ -1,44 +1,48 @@
...
+    class_attribute :add_caching_support
+    self.add_caching_support = true
+    class_attribute :attribute_names
+    self.attribute_names = []
 
     def self.attributes(*names)
-      attr_accessor(*names)
+      self.attribute_names |= names.map(&:to_sym)
+      # Silence redefinition of methods warnings
+      ActiveModelSerializers.silence_warnings do
```
Comment (groyoh) [inline lib/active_model_serializers/model.rb]: Why is this actually needed? Which method redefinition are you actually trying to silence?

Wouldn't

```ruby
names.map!(&to_sym)
attr_accessor(names - self.attribute_name)
self.attribute_names |= names
```

work instead of messing up with $VERBOSE?

File: test/action_controller/serialization_test.rb (hunk 5 above)
Comment (bf4) [inline test/action_controller/serialization_test.rb]: well, caught it because it raised an error, undefined attribute.. makes me wonder if it was being used...

File: test/fixtures/poro.rb (hunk 22 above)
Comment (bf4) [inline test/fixtures/poro.rb]: It makes more sense internally... we could deprecate it, but first look through issues and google group for anyone using it

File: lib/active_model_serializers/model.rb (hunk 23 above)
Comment (bf4) [inline lib/active_model_serializers/model.rb]: the `attr_accessor`.  It still yields warnings when, e.g. redefining `updated_at`, which isn't an `attribute` on the model by default

File: lib/active_model_serializers/model/caching.rb (hunk 13 above)
Comment (bf4) [inline lib/active_model_serializers/model/caching.rb]: note that `updated_at` isn't included in `attribute_names` and has no defined `updated_at=` setter.

File: lib/active_model_serializers/model/caching.rb [hunk 24]
```diff
@@ -0,0 +1,25 @@
+module ActiveModelSerializers
+  class Model
+    module Caching
+      extend ActiveSupport::Concern
+
+      included do
+        attr_writer :updated_at
+        attributes :id
```
Comment (bf4) [inline lib/active_model_serializers/model/caching.rb]: `id` will always be in `attributes`

File: test/action_controller/adapter_selector_test.rb (hunk 3 above)
Comment (bf4) [inline test/action_controller/adapter_selector_test.rb]: ref https://github.com/rails-api/active_model_serializers/pull/1984/files#r89365145

**Comment by groyoh** [comment]
LGTM! Any more work to be done for this?

**Comment by bf4** [comment]
@groyoh Just looks for consensus since there are some important changes in here.

**Comment by bf4** [comment]
@rails-api/ams @rails-api/commit ready for final review.  Should I break this up? Good enough?

File: lib/active_model_serializers/model/caching.rb [hunk 25]
```diff
@@ -4,19 +4,26 @@ module Caching
...
+      # there are other ways of setting the +cache_key+ a serializer uses.
       def cache_key
         ActiveSupport::Cache.expand_cache_key([
           self.class.model_name.name.downcase,
           "#{id}-#{updated_at.strftime('%Y%m%d%H%M%S%9N')}"
         ].compact)
       end
 
-      # Defaults to the time the serializer file was modified.
+      # When no set, defaults to the time the file was modified.
```
Comment (groyoh) [inline lib/active_model_serializers/model/caching.rb]: `s/no/not/` ?

**Comment by groyoh** [comment]
Isn't there too many changes here?

**Comment by bf4** [comment]
@groyoh 
> Isn't there too many changes here?

Yes, but I'm not sure how I'd like to break it up, since they go well together... part of the issue is that making the test attributes explicit, we need to decide when `id` or `updated_at` should be in the attributes hash, IIRC

I'm open to any suggestions how best to do it. That's why I'm not comfortable merging just yet.

File: test/serializers/associations_test.rb [hunk 26]
```diff
@@ -62,7 +62,13 @@ def test_serializer_options_are_passed_into_associations_serializers
...
                       .detect { |assoc| assoc.key == :comments }
 
-        assert association.serializer.first.custom_options[:custom_options]
+        comment_serializer = association.serializer.first
+        class << comment_serializer
+          def custom_options
+            instance_options
+          end
+        end
+        assert comment_serializer.custom_options
```
Comment (bf4) [inline test/serializers/associations_test.rb]: thanks, you're right! I lost the `custom_options[:custom_options]`

**Comment by bf4** [comment]
@rails-api/ams @rails-api/commit ready for review again

File: test/action_controller/json_api/transform_test.rb [hunk 27]
```diff
@@ -5,9 +5,17 @@ module Serialization
     class JsonApi
       class KeyTransformTest < ActionController::TestCase
         class KeyTransformTestController < ActionController::Base
-          class Post < ::Model; end
-          class Author < ::Model; end
-          class TopComment < ::Model; end
+          class Post < ::Model
+            attributes :title, :body, :publish_at
+            associations :author, :top_comments
```
Comment (NullVoxPopuli) [inline test/action_controller/json_api/transform_test.rb]: i'm certainly a fan of this being more explicit

File: test/active_model_serializers/model_test.rb [hunk 28]
```diff
@@ -18,5 +18,52 @@ def test_initialization_with_string_keys
 
       assert_equal model_instance.read_attribute_for_serialization(:key), value
     end
+
+    def test_attributes_can_be_read_for_serialization
```
Comment (NullVoxPopuli) [inline test/active_model_serializers/model_test.rb]: should new tests be using `test '...' do`?

File: test/cache_test.rb (hunk 16 above)
Comment (NullVoxPopuli) [inline test/cache_test.rb]: :+1:

File: test/fixtures/poro.rb (hunk 22 above)
Comment (NullVoxPopuli) [inline test/fixtures/poro.rb]: depracte in another PR or this one while we're talkin' about it?

**Comment by NullVoxPopuli** [comment]
I'm mostly :+1:

File: test/active_model_serializers/model_test.rb (hunk 28 above)
Comment (bf4) [inline test/active_model_serializers/model_test.rb]: oh, I guess so.  I guess I haven't completely switch to activesupport form.  I mostly just wanted the stable interface vs. minitest

File: test/fixtures/poro.rb (hunk 22 above)
Comment (bf4) [inline test/fixtures/poro.rb]: @NullVoxPopuli maybe we can make a project and put this in the queue as a 'note'?

Comment (NullVoxPopuli) [inline test/fixtures/poro.rb]: I dig it
